server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -230,6 +344,142 @@ private void abortMergeTask(MergeTask mergeTask) {
+                return;
+            }
+            for (NodeEnvironment.DataPath dataPath : dataPaths) {
+                try {
+                    FsInfo.Path fsInfo = getFSInfo(dataPath); // uncached
+                    if (leastAvailablePath == null || leastAvailablePath.getAvailable().getBytes() > fsInfo.getAvailable().getBytes()) {
+                        leastAvailablePath = fsInfo;
+                    }
+                } catch (IOException e) {
+                    if (fsInfoException == null) {
+                        fsInfoException = e;
+                    } else {
+                        fsInfoException.addSuppressed(e);
+                    }
+                }
+            }
+            if (fsInfoException != null) {
+                LOGGER.warn("unexpected exception reading filesystem info", fsInfoException);
+            }
+            if (leastAvailablePath == null) {
+                LOGGER.error("Cannot read filesystem info");
+                return;
+            }
+            // subtract disk space that already running merges are expected to fill
+            long leastAvailableDiskSpaceBytes = leastAvailablePath.getAvailable().getBytes();
+            for (MergeTask mergeTask : runningMergeTasks) {
+                leastAvailableDiskSpaceBytes -= mergeTask.estimatedRemainingMergeSize();
+            }
+            // also subtract the configured free disk space threshold
+            leastAvailableDiskSpaceBytes -= getFreeBytesThreshold(leastAvailablePath.getTotal(), highStageWatermark, highStageMaxHeadroom)
+                .getBytes();
+            // the rest is the maximum disk space available for a new merge task
+            long maxMergeSizeLimit = Math.max(0L, leastAvailableDiskSpaceBytes);
+            queuedMergeTasks.updateMaxPriorityLimit(maxMergeSizeLimit);
+        }
+
+        private static ByteSizeValue getFreeBytesThreshold(
+            ByteSizeValue total,
+            RelativeByteSizeValue watermark,
+            ByteSizeValue maxHeadroom
+        ) {
+            // If bytes are given, they can be readily returned as free bytes. If percentages are given, we need to calculate the free
+            // bytes.
+            if (watermark.isAbsolute()) {
+                return watermark.getAbsolute();
+            }
+            return ByteSizeValue.subtract(total, watermark.calculateValue(total, maxHeadroom));
+        }
+    }
+
+    static class PriorityBlockingQueueWithMaxLimit<E> {
+        private final ToLongFunction<? super E> priorityFunction;
+        private final PriorityQueue<E> priorityQueue;
+        private final ReentrantLock lock;
+        private final Condition elementAvailable;
+        private long maxPriorityLimit;
+
+        PriorityBlockingQueueWithMaxLimit(ToLongFunction<? super E> priorityFunction, long maxPriorityLimit) {
+            this.priorityFunction = priorityFunction;
```

Comment:
This function and it's name seems slightly confusing. I think it would be simpler to have it be a predicate that says whether it can run or not? We'd still need the wakeup when disk usage monitoring returns but that could just be a wakeup function.

That way it can also handle all criteria, like heap, disk etc. that we come up with in the future.

In fact, it should maybe reserve the capacity too, such that we conservatively ensure we do not run out. We can add that later, no need for it initially (and maybe it is good enough without it).

Reply:
> I think it would be simpler to have it be a predicate that says whether it can run or not? We'd still need the wakeup when disk usage monitoring returns but that could just be a wakeup function.

Ok, I think that would alright.

> That way it can also handle all criteria, like heap, disk etc. that we come up with in the future.

Mhm, I'm not sure about that. It might be the case that a small disk-wise merge is a large heap-wise merge...
The reason it's OK to block at the priority queue's head for disk-space reasons is because any other elements in the queue would need to block as well, since they are larger. But if the priority queue is still ordered by disk space but sometimes we need to defer by heap size, it won't work.

Reply:
Sure. I can live with that for now. It means that if a merge is blocked by heap (which will be vectors only) we can still accept a newer merge that is not for vectors and execute it. We can then see if we ever need to fix that. I think it is ok.